test/test_base.py

```diff
@@ -17,47 +19,66 @@ def __init__(self):
         self.network_name = self.tmpdir.name
         self.logfile = None
         self.server = None
+        self.network = None
 
         atexit.register(self.cleanup)
 
         print(f"\nWarnet test base started")
 
-
-    def cleanup(self, signum = None, frame = None):
-        if self.server is None:
-            return
-
+    def _cleanup_docker(self):
         try:
-            print("\nStopping network")
-            self.warcli("network down")
-
-            self.wait_for_all_tanks_status(target="none", timeout=60, interval=1)
-
-            print("\nStopping server")
-            self.warcli("stop", False)
-        except Exception as e:
-            # Remove the temporary docker network when we quit.
-            # If the warnet server exited prematurely then docker-compose down
-            # likely did not succeed or was never executed.
-            print(f"Error stopping server: {e}")
             print("Attempting to cleanup docker network")
             wn = Warnet.from_network(self.network_name)
             wn.docker_compose_down()
+        except FileNotFoundError:
+            pass
+        except Exception:
+            pass
+
+    def _cleanup_network(self):
+        if self.network is not None:
+            try:
+                print("\nStopping network")
+                self.warcli("network down")
+
+                self.wait_for_all_tanks_status(target="none", timeout=60, interval=1)
+
+                print("\nStopping server")
+                self.warcli("stop", False)
+            except Exception as e:
+                # Remove the temporary docker network when we quit.
+                # If the warnet server exited prematurely then docker-compose down
+                # likely did not succeed or was never executed.
+                print(f"Error stopping server: {e}")
+                self._cleanup_docker()
+
+    def _cleanup_server(self):
+        # kill the server process by its PID if it still survives
+        if self.server:
+            print(f"killing server with pid {self.server.pid}")
+            os.killpg(os.getpgid(self.server.pid), signal.SIGTERM)
+
+    def cleanup(self, signum = None, frame = None):
+        if self.server is None:
+            return
 
-        print("\nRemaining server output:")
-        print(self.logfile.read())
-        self.logfile.close()
+        self._cleanup_network()
+        self._cleanup_server()
 
-        self.logfile = None
-        self.server.terminate()
-        self.server = None
+        if self.logfile:
```

Comment:
Don't 100% understand why this can be None, but it was often for me?

Reply:
Sometimes it gets called twice because of atexit. I want to garuntee it always gets run at least once when we quit